File: tests/test_temporal_erh.py
"""
Unit tests for temporal ERH module.
"""

import sys
import os
from pathlib import Path

# Add simulation to path
simulation_dir = Path(__file__).parent.parent / "simulation"
if str(simulation_dir) not in sys.path:
    sys.path.insert(0, str(simulation_dir))

# Also add project root for absolute imports
project_root = Path(__file__).parent.parent
if str(project_root) not in sys.path:
    sys.path.insert(0, str(project_root))

import numpy as np
import pytest

# Import with error handling
try:
    from core.action_space import generate_world, Action
    from core.judgement_system import BiasedJudge
    from core.temporal_erh import (
        compute_Pi_temporal,
        compute_E_temporal,
        compute_baseline_temporal,
        track_error_evolution,
        simulate_mule_effect,
        detect_mule_anomalies
    )
except ImportError as e:
    pytest.skip(f"Failed to import required modules: {e}", allow_module_level=True)


class TestTemporalERH:
    """Test temporal ERH functions."""
    
    def test_compute_Pi_temporal(self):
        """Test Pi(x,t) computation."""
        # Create mock primes history
        primes_history = []
        for t in range(3):
            primes = [
                Action(id=i, c=10 + i*5, V=0.5, w=1.0) for i in range(5)
            ]
            primes_history.append(primes)
        
        Pi_xt = compute_Pi_temporal(primes_history, time_steps=3, X_max=50)
        
        assert Pi_xt.shape == (3, 50)
        assert np.all(Pi_xt >= 0)
        assert Pi_xt[0, 9] == 0  # No primes with c <= 10 at t=0
        assert Pi_xt[0, 14] >= 1  # At least one prime with c <= 15
    
    def test_compute_baseline_temporal(self):
        """Test baseline computation."""
        B_xt = compute_baseline_temporal(
            time_steps=5, X_max=20, baseline_type='prime_theorem'
        )
        
        assert B_xt.shape == (5, 20)
        assert np.all(B_xt >= 0)
    
    def test_compute_E_temporal(self):
        """Test E(x,t) computation."""
        Pi_xt = np.array([[1, 2, 3], [2, 3, 4]])
        B_xt = np.array([[0.5, 1.0, 1.5], [1.0, 1.5, 2.0]])
        
        E_xt = compute_E_temporal(Pi_xt, B_xt, time_steps=2, X_max=3)
        
        assert E_xt.shape == (2, 3)
        expected = Pi_xt - B_xt
        np.testing.assert_array_almost_equal(E_xt, expected)
    
    def test_track_error_evolution(self):
        """Test error evolution tracking."""
        actions_history = []
        judge = BiasedJudge(bias_strength=0.2)
        
        for t in range(3):
            actions = generate_world(num_actions=100, random_seed=42 + t)
            actions_history.append(actions)
        
        results = track_error_evolution(
            actions_history, judge, tau=0.3, time_steps=3, X_max=50
        )
        
        assert 'Pi_xt' in results
        assert 'B_xt' in results
        assert 'E_xt' in results
        assert results['Pi_xt'].shape == (3, 50)
        assert results['E_xt'].shape == (3, 50)
    
    def test_simulate_mule_effect(self):
        """Test Mule effect simulation."""
        E_xt = np.zeros((5, 10))
        E_xt[2, 5] = 1.0  # Initial error
        
        E_xt_mule = simulate_mule_effect(
            E_xt, mule_time=2, mule_strength=2.0, X_max=10
        )
        
        assert E_xt_mule.shape == E_xt.shape
        # Mule effect should modify the error
        assert not np.array_equal(E_xt, E_xt_mule)
    
    def test_detect_mule_anomalies(self):
        """Test anomaly detection."""
        # Create error array with violations
        E_xt = np.zeros((3, 10))
        E_xt[1, 5] = 5.0  # Large violation
        
        anomalies = detect_mule_anomalies(E_xt, C=1.0, epsilon=0.1, X_max=10)
        
        assert len(anomalies) > 0
        assert any(a['time'] == 1 and a['complexity'] == 5 for a in anomalies)


if __name__ == '__main__':
    pytest.main([__file__, '-v'])

File: julia/test/test_temporal_erh.jl
"""
Tests for TemporalERH.jl — covers Π(x,t), B(x,t), E(x,t) computation,
baseline variants, track_error_evolution, Mule effect, anomaly detection,
and stochastic perturbation.
"""

@testset "TemporalERH" begin

    import ..TemporalERH as TERH
    using Random: Xoshiro

    # Convenience: build a TAction-like struct compatible with the module
    struct TAct
        id::Int
        c::Float64
        V::Float64
        w::Float64
    end

    # -----------------------------------------------------------------------
    # 1. compute_Pi_temporal
    # -----------------------------------------------------------------------
    @testset "compute_Pi_temporal" begin
        T     = 3
        X_max = 10

        # t=1: one prime at c=5
        # t=2: two primes at c=3 and c=8
        # t=3: empty
        primes_history = [
            [TAct(1, 5.0, 0.5, 1.0)],
            [TAct(2, 3.0, 0.4, 1.0), TAct(3, 8.0, 0.6, 1.0)],
            TAct[]
        ]

        Pi_xt = TERH.compute_Pi_temporal(primes_history, T, X_max)
        @test size(Pi_xt) == (T, X_max)
        @test Pi_xt[1, 5]  == 1.0   # one prime with c ≤ 5 at t=1
        @test Pi_xt[1, 4]  == 0.0   # no prime with c ≤ 4 at t=1
        @test Pi_xt[2, 8]  == 2.0   # two primes with c ≤ 8 at t=2
        @test Pi_xt[2, 2]  == 0.0   # no prime with c ≤ 2 at t=2
        @test Pi_xt[3, 10] == 0.0   # empty at t=3
    end

    # -----------------------------------------------------------------------
    # 2. compute_baseline_temporal — all variants
    # -----------------------------------------------------------------------
    @testset "compute_baseline_temporal" begin
        T, X = 4, 20
        for bt in ("prime_theorem", "linear", "logarithmic_integral", "power_law")
            B = TERH.compute_baseline_temporal(T, X; baseline_type=bt)
            @test size(B) == (T, X)
            @test all(b -> b >= 0, B)
        end

        # Time-dependent drift increases values over time
        B_static = TERH.compute_baseline_temporal(10, 20; time_dependent=false)
        B_drift  = TERH.compute_baseline_temporal(10, 20; time_dependent=true)
        @test B_drift[10, 10] >= B_drift[1, 10]   # later times have higher baseline
    end

    # -----------------------------------------------------------------------
    # 3. compute_E_temporal
    # -----------------------------------------------------------------------
    @testset "compute_E_temporal" begin
        T, X = 5, 15
        Pi = ones(Float64, T, X)
        B  = 0.5 .* ones(Float64, T, X)
        E  = TERH.compute_E_temporal(Pi, B, T, X)
        @test size(E) == (T, X)
        @test all(e -> e ≈ 0.5, E)

        # Zero Pi → negative error
        E2 = TERH.compute_E_temporal(zeros(T, X), B, T, X)
        @test all(e -> e ≈ -0.5, E2)
    end

    # -----------------------------------------------------------------------
    # 4. track_error_evolution
    # -----------------------------------------------------------------------
    @testset "track_error_evolution" begin
        T     = 4
        X_max = 15
        # Build synthetic actions
        function make_actions(n, seed)
            rng = Xoshiro(seed)
            [TAct(i, 1.0 + 99rand(rng), rand(rng), 0.5 + rand(rng)) for i in 1:n]
        end
        actions_history = [make_actions(50, 10+t) for t in 1:T]

        # Judge: error if V < 0.4
        judge_fn(a, τ) = a.V < τ

        result = TERH.track_error_evolution(
            actions_history, judge_fn; τ=0.4, time_steps=T, X_max=X_max
        )
        @test haskey(result, "Pi_xt");  @test size(result["Pi_xt"]) == (T, X_max)
        @test haskey(result, "B_xt");   @test size(result["B_xt"])  == (T, X_max)
        @test haskey(result, "E_xt");   @test size(result["E_xt"])  == (T, X_max)
        @test haskey(result, "primes_history")
        @test length(result["primes_history"]) == T
    end

    # -----------------------------------------------------------------------
    # 5. simulate_mule_effect
    # -----------------------------------------------------------------------
    @testset "simulate_mule_effect" begin
        T, X = 6, 20
        E_xt = zeros(Float64, T, X)
        # Set some non-zero baseline so the mule multiplier has something to amplify
        E_xt[3, :] .= 2.0

        E_mod = TERH.simulate_mule_effect(E_xt, 3; mule_strength=3.0,
                                           mule_complexity_range=(8, 15), seed=5)
        @test size(E_mod) == (T, X)
        # The mule time row should differ from the original
        @test !all(E_mod[3, :] .== E_xt[3, :])
        # Earlier rows should be unchanged
        @test E_mod[1, :] == E_xt[1, :]
        @test E_mod[2, :] == E_xt[2, :]

        # mule_time beyond time_steps → no modification
        E_noop = TERH.simulate_mule_effect(E_xt, 99; mule_strength=5.0)
        @test E_noop == E_xt
    end

    # -----------------------------------------------------------------------
    # 6. detect_mule_anomalies
    # -----------------------------------------------------------------------
    @testset "detect_mule_anomalies" begin
        T, X = 5, 10
        E_small = 0.01 .* ones(Float64, T, X)
        no_anomalies = TERH.detect_mule_anomalies(E_small; C=1.0, epsilon=0.1)
        @test isempty(no_anomalies)

        # Inject a large violation at (t=2, x=5)
        E_big = copy(E_small)
        E_big[2, 5] = 1000.0
        anomalies = TERH.detect_mule_anomalies(E_big; C=1.0, epsilon=0.1,
                                                threshold_multiplier=1.5, X_max=X)
        @test length(anomalies) >= 1
        found = filter(a -> a["complexity"] == 5, anomalies)
        @test !isempty(found)
        @test found[1]["violation_ratio"] > 1.5

        # All anomaly dicts have required keys
        for a in anomalies
            @test haskey(a, "time")
            @test haskey(a, "complexity")
            @test haskey(a, "error_magnitude")
            @test haskey(a, "erh_bound")
        end
    end

    # -----------------------------------------------------------------------
    # 7. add_stochastic_perturbation
    # -----------------------------------------------------------------------
    @testset "add_stochastic_perturbation" begin
        T, X = 10, 20
        E_base = zeros(Float64, T, X)
        E_pert = TERH.add_stochastic_perturbation(E_base; noise_scale=0.1,
                                                   correlation_time=2, seed=7)
        @test size(E_pert) == (T, X)
        # Perturbed should differ from original
        @test !all(E_pert .== E_base)
        # Standard deviation of noise ≈ noise_scale (rough check)
        @test std(E_pert) < 0.5   # not unbounded

        # Different seeds give different results
        E2 = TERH.add_stochastic_perturbation(E_base; noise_scale=0.1, seed=99)
        @test E_pert != E2
    end

    # -----------------------------------------------------------------------
    # 8. Round-trip: Pi → B → E consistency
    # -----------------------------------------------------------------------
    @testset "Pi-B-E round-trip" begin
        T, X = 3, 8
        primes_history = [
            [TAct(i, Float64(i), 0.5, 1.0) for i in 1:k]
            for k in [3, 5, 7]
        ]
        Pi = TERH.compute_Pi_temporal(primes_history, T, X)
        B  = TERH.compute_baseline_temporal(T, X)
        E  = TERH.compute_E_temporal(Pi, B, T, X)
        @test all(isfinite, E)
        @test E == Pi .- B
    end

end # @testset "TemporalERH"
